server/src/main/java/org/opensearch/action/ActionModule.java

```diff
@@ -954,13 +967,86 @@ protected void configure() {
                 bind(supportAction).asEagerSingleton();
             }
         }
+
+        // register dynamic ActionType -> transportAction Map used by NodeClient
+        bind(DynamicActionRegistry.class).toInstance(dynamicActionRegistry);
     }
 
     public ActionFilters getActionFilters() {
         return actionFilters;
     }
 
+    public DynamicActionRegistry getDynamicActionRegistry() {
+        return dynamicActionRegistry;
+    }
+
     public RestController getRestController() {
         return restController;
     }
+
+    /**
+     * The DynamicActionRegistry maintains a registry mapping {@link ActionType} instances to {@link TransportAction} instances.
+     * <p>
+     * This class is modeled after {@link NamedRegistry} but provides both register and unregister capabilities.
+     *
+     * @opensearch.internal
+     */
+    public static class DynamicActionRegistry {
+        // This is the unmodifiable actions map created during node bootstrap, which
+        // will continue to link ActionType and TransportAction pairs from core and plugin
+        // action handler registration.
+        private Map<ActionType, TransportAction> actions = Collections.emptyMap();
+        // A dynamic registry to add or remove ActionType / TransportAction pairs
+        // at times other than node bootstrap.
+        private final Map<ActionType<?>, TransportAction<?, ?>> registry = new ConcurrentHashMap<>();
+
+        /**
+         * Register the immutable actions in the registry.
+         *
+         * @param actions The injected map of {@link ActionType} to {@link TransportAction}
+         */
+        public void registerUnmodifiableActionMap(Map<ActionType, TransportAction> actions) {
```

Comment:
Trying to understand, we could've directly passed in the map L516 when `DynamicActionRegistry` is created. Did you want to keep the structure of the injected information ?

Reply:
> Trying to understand, we could've directly passed in the map L516 when `DynamicActionRegistry` is created. 

Yep but it's a different map. 

Given `getActions()` this was the simplest approach and the first thing I tried.

The bits we need are the ActionType key and TransportAction singleton instance value.

The signature of `actions` is `Map<String, ActionHandler<?, ?>>` where the ActionHandler does have the ActionType key, but only the _class_ of the TransportAction.  And the only way to fetch that singleton instance that was injected is to pass around the injector (or a map retrieved from the injector, as was done to send the "map we need" to NodeClient and is now inside the DynamicActionRegistry).

> Did you want to keep the structure of the injected information ?

My original implementation kept this map injected as-is and handled the two different maps in the NodeClient.  However, based on initial review it was requested I keep those details out of the client.   

But we still need to provide an ActionType instance key (easy) and fetch a singleton TransportAction value that we can run the `doExecute()` method on.  

We *could* use the class and get the right instance if we keep a copy of the injector inside the registry, and in theory that might help out with the concern over rawtypes.

But that's also new code/patterns and I really tried to keep this as close to the way it already was as I could.